csharp/src/ProtocolBuffers.Test/GeneratedMessageTest.cs

```diff
@@ -510,5 +536,221 @@ public void EmptyPackedValue()
             TestPackedTypes empty = new TestPackedTypes.Builder().Build();
             Assert.AreEqual(0, empty.SerializedSize);
         }
+
+        // oneof tests
+        [Test]
+        public void TestOneofEmumCase()
+        {
+            TestOneof2 message = TestOneof2.CreateBuilder()
+                .SetFooInt(123).SetFooString("foo").SetFooCord("bar").Build();
+            TestUtil.AssertAtMostOneFieldSetOneof(message);
+        }
+
+        [Test]
+        public void TestClearOneof()
+        {
+            TestOneof2.Builder builder = TestOneof2.CreateBuilder().SetFooInt(123);
+            Assert.AreEqual(TestOneof2.FooOneofCase.FooInt, builder.FooCase);
+            builder.ClearFoo();
+            Assert.AreEqual(TestOneof2.FooOneofCase.FooNotSet, builder.FooCase);
+        }
+
+        [Test]
+        public void TestSetOneofClearsOthers()
+        {
+            TestOneof2.Builder builder = TestOneof2.CreateBuilder();
+            TestOneof2 message =
+                builder.SetFooInt(123).SetFooString("foo").BuildPartial();
+            Assert.IsTrue(message.HasFooString);
+            TestUtil.AssertAtMostOneFieldSetOneof(message);
+
+            message = builder.SetFooCord("bar").BuildPartial();
+            Assert.IsTrue(message.HasFooCord);
+            TestUtil.AssertAtMostOneFieldSetOneof(message);
+
+            message = builder.SetFooStringPiece("baz").BuildPartial();
+            Assert.IsTrue(message.HasFooStringPiece);
+            TestUtil.AssertAtMostOneFieldSetOneof(message);
+
+            message = builder.SetFooBytes(TestUtil.ToBytes("qux")).BuildPartial();
+            Assert.IsTrue(message.HasFooBytes);
+            TestUtil.AssertAtMostOneFieldSetOneof(message);
+
+            message = builder.SetFooEnum(TestOneof2.Types.NestedEnum.FOO).BuildPartial();
+            Assert.IsTrue(message.HasFooEnum);
+            TestUtil.AssertAtMostOneFieldSetOneof(message);
+
+            message = builder.SetFooMessage(
+                TestOneof2.Types.NestedMessage.CreateBuilder().SetQuxInt(234).Build()).BuildPartial();
+            Assert.IsTrue(message.HasFooMessage);
+            TestUtil.AssertAtMostOneFieldSetOneof(message);
+
+            message = builder.SetFooInt(123).BuildPartial();
+            Assert.IsTrue(message.HasFooInt);
+            TestUtil.AssertAtMostOneFieldSetOneof(message);
+        }
+
+        [Test]
+        public void TestOneofTypes()

```

Comment:
Any reason for just using blocks instead of multiple tests here? I'd have TestOneofTypes_Primitive, TestOneofTypes_Enum etc.